### PR TITLE
Add new practical nearest neighbor search guide

### PR DESCRIPTION
**Pull request by jobergum**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: en/performance/practical-nearest-neighbor-search.md [hunk 1]
```diff
@@ -0,0 +1,1945 @@
...
+                    "title": "Nothing's Gonna Change My Love For You",
+                    "artist": "Glenn Medeiros"
+                }
+            }
+        ]
+    }
+}
+
+{% endhighlight %}
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:690]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 2]
```diff
@@ -0,0 +1,1945 @@
...
+  --verbose --stdin --endpoint http://localhost:8080
+</pre>
+</div>
+
+The following query examples uses a static query vector embedding for the
+query string *Total Eclipse Of The Heart*. The query embedding was obtained by the
+following snippet using [sentence-transformers](https://www.sbert.net/):
+
+<pre>
+{% highlight python%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:724]:
```suggestion
<pre>{% highlight python%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 3]
```diff
@@ -0,0 +1,1945 @@
...
+<pre data-test="exec" data-test-assert-contains="Bonnie Tyler">
+$ vespa query \
+    'yql=select title, artist from track where {approximate:false,targetHits:10}nearestNeighbor(embedding,q)' \
+    'hits=1' \
+    'ranking=closeness' \
+    "ranking.features.query(q)=$Q"
+</pre>
+</div>
+
+The query expresses the following :
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:748]:
```suggestion
Query breakdown:
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 4]
```diff
@@ -0,0 +1,1945 @@
...
+in the `rank-profile`.
+Not specifying [ranking](../reference/query-api-reference.html#ranking.profile) will cause
+Vespa to use [nativeRank](../nativerank.html) which does not use the vector similarity, causing
+results to be randomly sorted. 
+
+The above exact nearest neighbor search will return the following
+[result](../reference/default-result-format.html):
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:766]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 5]
```diff
@@ -0,0 +1,1945 @@
...
+                "fields": {
+                    "title": "Total Eclipse Of The Heart",
+                    "artist": "Bonnie Tyler"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %}  
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:801]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 6]
```diff
@@ -0,0 +1,1945 @@
...
+The exact search takes approximately 51ms, performing 95,666 distance calculations. 
+A total of about 120 documents were exposed to the first-phase ranking during the search as can be seen from
+`totalCount`.  Vespa's exact nearest neighbor search uses chunked vector distance calculations, splitting
+the vectors into chunks, the chunked distance calculates reduces the computational complexity. Onl
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:805]: Onl ?

File: en/performance/practical-nearest-neighbor-search.md [hunk 7]
```diff
@@ -0,0 +1,1945 @@
...
+    'ranking=closeness-t4' \
+    "ranking.features.query(q)=$Q"
+</pre>
+</div>
+
+Now, the exact search latency is reduced by using more threads, 
+see [multi-threaded searching and ranking](practical-search-performance-guide.html#multi-threaded-search-and-ranking)
+for more on this topic.
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:825]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 8]
```diff
@@ -0,0 +1,1945 @@
...
+{
+    "timing": {
+        "querytime": 0.019,
+        "summaryfetchtime": 0.001,
+        "searchtime": 0.021
+    }
+}  
+{% endhighlight %}  
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:834]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 9]
```diff
@@ -0,0 +1,1945 @@
...
+    'hits=1' \
+    'ranking=closeness' \
+    "ranking.features.query(q)=$Q"
+</pre>
+</div>
+
+Which returns the following response:
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:860]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 10]
```diff
@@ -0,0 +1,1945 @@
...
+                "fields": {
+                    "title": "Total Eclipse Of The Heart",
+                    "artist": "Bonnie Tyler"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:895]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 11]
```diff
@@ -0,0 +1,1945 @@
...
+
+Now, the query is significantly faster, but also uses less resources during the search. To get latency down
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:897]:
```suggestion
Now, the query is significantly faster, and also uses less resources during the search. To get latency down 
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 12]
```diff
@@ -0,0 +1,1945 @@
...
+    'yql=select title, artist from track where {targetHits:10}nearestNeighbor(embedding,q) and title contains "heart"' \
+    'hits=2' \
+    'ranking=closeness-t4' \
+    "ranking.features.query(q)=$Q"
+</pre>
+</div>
+
+Which returns the following response:
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:936]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 13]
```diff
@@ -0,0 +1,1945 @@
...
+                "fields": {
+                    "title": "Heart Of My Heart",
+                    "artist": "Quest"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:980]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 14]
```diff
@@ -0,0 +1,1945 @@
...
+    'hits=2' \
+    'ranking=closeness-t4' \
+    "ranking.features.query(q)=$Q"
+</pre>
+</div>
+
+In the previous examples, since the rank-profile did only use the `closeness` rank feature,  
+the matching would not impact the score anyway. 
+
+Vespa also allow combining the `nearestNeighbor` query operator with any other Vespa query operator.
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1031]:
```suggestion
Vespa also allows combining the `nearestNeighbor` query operator with any other Vespa query operator.  
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 15]
```diff
@@ -0,0 +1,1945 @@
...
+    'hits=2' \
+    'ranking=closeness-t4' \
+    "ranking.features.query(q)=$Q"
+</pre>
+</div>
+
+The above query returns 
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1086]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 16]
```diff
@@ -0,0 +1,1945 @@
...
+                    "title": "Closer To The Heart",
+                    "artist": "Rush",
+                    "popularity": 100
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1138]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 17]
```diff
@@ -0,0 +1,1945 @@
...
+$ vespa query \
+    'yql=select matchfeatures, title, popularity, artist from track where {distanceThreshold:0.7,targetHits:10}nearestNeighbor(embedding,q) and popularity > 80' \
+    'hits=2' \
+    'ranking=closeness-t4' \
+    "ranking.features.query(q)=$Q"
+</pre>
+</div>
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1155]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 18]
```diff
@@ -0,0 +1,1945 @@
...
+                    "title": "Total Eclipse Of The Heart",
+                    "artist": "Bonnie Tyler",
+                    "popularity": 100
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1194]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 19]
```diff
@@ -0,0 +1,1945 @@
...
+            bm25(title)
+            closeness(field, embedding)
+        }
+    }
+</pre>
+
+The query returns the following result:
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1268]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 20]
```diff
@@ -0,0 +1,1945 @@
...
+                        "rawScore(tags)": 0.0
+                    },
+                    "title": "Eclipse",
+                    "artist": "Kyoto Jazz Massive"
+                }
+            }
+        ]
+    }
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1323]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 21]
```diff
@@ -0,0 +1,1945 @@
...
+    'ranking=hybrid' \
+    "ranking.features.query(q)=$Q" \
+    'ranking.features.query(wVector)=40'
+</pre>
+</div>
+
+Which changes the order and a different hit is surfaced at position two:
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1349]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 22]
```diff
@@ -0,0 +1,1945 @@
...
+                    },
+                    "title": "Closer To The Heart",
+                    "artist": "Rush"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1405]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 23]
```diff
@@ -0,0 +1,1945 @@
...
+    'ranking.features.query(wVector)=340' \
+    'userProfile={"love songs":1, "love":1,"80s":1}' 
+</pre>
+</div>
+
+In this case, another document is surfaced at position 2, which have a non-zero personalized score,
+notice that `totalCount` increases as the `wand` query operator brought more hits into `first-phase` ranking.
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1434]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 24]
```diff
@@ -0,0 +1,1945 @@
...
+                    },
+                    "title": "Nothing's Gonna Change My Love For You",
+                    "artist": "Glenn Medeiros"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1490]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 25]
```diff
@@ -0,0 +1,1945 @@
...
+</div>
+
+This query returns 100 documents, since only the first operand of the `rank` query operator was used for 
+*retrieval*, the sparse `userQuery()` representation was only used to calculate sparse 
+[rank features](../reference/rank-features.html) for
+the results retrieved by the `nearestNeighbor`. Sparse rank features such as `bm25(title)` for example.
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1582]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 26]
```diff
@@ -0,0 +1,1945 @@
...
+                    },
+                    "title": "Closer To The Heart",
+                    "artist": "Rush"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1638]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 27]
```diff
@@ -0,0 +1,1945 @@
...
+[Vespa passage ranking](https://github.com/vespa-engine/sample-apps/blob/master/msmarco-ranking/passage-ranking.md)
+for complete examples of different retrieval strategies for multi-phase ranking funnels.
+
+## Multiple nearest neighbor search operators in the same query 
+This section looks at how to use multiple `nearestNeighbor` query operator instances in the same Vespa query request. 
+
+First, the query embedding for *Total Eclipse Of The Heart*:
+
+<pre>
+{% highlight python%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1670]:
```suggestion
<pre>{% highlight python%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 28]
```diff
@@ -0,0 +1,1945 @@
...
+from sentence_transformers import SentenceTransformer
+model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
+print(model.encode("Total Eclipse Of The Heart").tolist())
+{% endhighlight %}
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1675]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 29]
```diff
@@ -0,0 +1,1945 @@
...
+
+Secondly, the query embedding for *Summer of '69*:
+<pre>
+{% highlight python%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1683]:
```suggestion
<pre>{% highlight python%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 30]
```diff
@@ -0,0 +1,1945 @@
...
+from sentence_transformers import SentenceTransformer
+model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
+print(model.encode("Summer of '69").tolist())
+{% endhighlight %}
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1688]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 31]
```diff
@@ -0,0 +1,1945 @@
...
+
+The following Vespa query combines two `nearestNeighbor` query operators 
+using logical disjunction (`OR`) and referencing two different
+query tensor inputs:
+
+- `ranking.features.query(q)` holding the **Total Eclipse Of The Heart* query vector.
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1699]:
```suggestion
- `ranking.features.query(q)` holding the *Total Eclipse Of The Heart* query vector.
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 32]
```diff
@@ -0,0 +1,1945 @@
...
+    'hits=2' \
+    'ranking=closeness-t4' \
+    "ranking.features.query(q)=$Q" \
+    "ranking.features.query(qa)=$QA" 
+</pre>
+</div>
+
+The above query returns 20 documents to first phase ranking, as seen from `totalCount`. Ten from each nearest neighbor query operator:
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1716]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 33]
```diff
@@ -0,0 +1,1945 @@
...
+                "source": "tracks",
+                "fields": {
+                    "title": "Summer of 69"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1758]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 34]
```diff
@@ -0,0 +1,1945 @@
...
+
+One can also use `label` annotation when there are multiple nearest neighbor search operators in the same query
+to differentiate which of them produced the match. 
+
+<div class="pre-parent">
+  <button class="d-icon d-duplicate pre-copy-button" onclick="copyPreContent(this)"></button>
+<pre data-test="exec" data-test-assert-contains='Total Eclipse Of The Heart'>
+$ vespa query \
+    'yql=select title, matchfeatures from track where ({ "label":"q", targetHits:10}nearestNeighbor(embedding,q)) or ({"label":"qa",targetHits:10}nearestNeighbor(embedding,qa))' \
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1767]:
```suggestion
    'yql=select title, matchfeatures from track where ({ label:"q", targetHits:10}nearestNeighbor(embedding,q)) or ({label:"qa",targetHits:10}nearestNeighbor(embedding,qa))' \
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 35]
```diff
@@ -0,0 +1,1945 @@
...
+</pre>
+</div>
+
+The above query annotates the two `nearestNeighbor` query operators using 
+[label](../reference/query-language-reference.html#label) query annotation. The result include 
+`match-features` so one can see which query operator retrieved the document from the 
+`closeness(label, ..)` feature output:
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1781]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 36]
```diff
@@ -0,0 +1,1945 @@
...
+                        "closeness(label,qa)": 0.5794361034642413
+                    },
+                    "title": "Summer of 69"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1831]:
```suggestion
{% endhighlight %}</pre>
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 37]
```diff
@@ -0,0 +1,1945 @@
...
+
+Note that the previous examples used `or` to combine the two operators. Using `and` instead, requires 
+that there are documents that is in both the top-k results. Increasing `targetHits` to 500,  
+finds 9 tracks that overlap. In this case both closeness labels have a non-zero score. 
+
+<div class="pre-parent">
+  <button class="d-icon d-duplicate pre-copy-button" onclick="copyPreContent(this)"></button>
+<pre data-test="exec" data-test-assert-contains='Summer Of Love'>
+$ vespa query \
+    'yql=select title, matchfeatures from track where ({ "label":"q", targetHits:500}nearestNeighbor(embedding,q)) and ({"label":"qa",targetHits:500}nearestNeighbor(embedding,qa))' \
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1841]:
```suggestion
    'yql=select title, matchfeatures from track where ({label:"q", targetHits:500}nearestNeighbor(embedding,q)) and ({label:"qa",targetHits:500}nearestNeighbor(embedding,qa))' \
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 38]
```diff
@@ -0,0 +1,1945 @@
...
+</pre>
+</div>
+
+Which returns the following top two hits. Note that the `closeness-label` rank profile
+uses `closeness(field, embedding)` which in the case of multiple nearest neighbor search operators 
+uses the maximum score to represent the unlabeled `closeness(field,embedding)`. This
+can be seen from the `relevance` value, compared with the labeled `closeness()` rank features. 
+
+<pre>
+{% highlight json%}
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1855]:
```suggestion
<pre>{% highlight json%}
```

File: en/performance/practical-nearest-neighbor-search.md [hunk 39]
```diff
@@ -0,0 +1,1945 @@
...
+                        "closeness(label,qa)": 0.47990179066646654
+                    },
+                    "title": "Midnight Heartache"
+                }
+            }
+        ]
+    }
+}
+{% endhighlight %} 
+</pre>
```
Comment (kkraune) [inline en/performance/practical-nearest-neighbor-search.md:1905]:
```suggestion
{% endhighlight %}</pre>
```

**Review by kkraune — APPROVED** [review]
Great guide! looking forward to trying it!

**Comment by jobergum** [comment]
I think this is fairly great, we might need to trickle more links to this guide throughput the doc guide? What you think @kkraune

**Comment by kkraune** [comment]
> 

Yes, we must link to this, and possibly remove some redundant content in the other guides :-)

**Pull request merged**